<a href="https://colab.research.google.com/github/mandlspr/ComplexLinearRegression/blob/main/Stats_Mini_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1 - Understand the data**

In [ ]:
# understand the tables and how they link with each other.
# discard any data we don’t find relevant.

# first, let's import our various tools in this notebook in order to perform our analysis and plottings.

In [ ]:
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px

For the sake of this study and due to a limited amount of time today, we will focus on ' ' and ' ' datasets. If time allows it, we will expand the analysis to ' ' dataset.

In [ ]:
df_orders_dataset = pd.read_csv('/content/olist_orders_dataset.csv')
df_orders_dataset

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 8 columns]

In [ ]:
df_orders_dataset.dtypes

# it is clear that we don't have the correct data types in this dataset
# we will have to perform some changes in the cleaning section (Dates)

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [ ]:
df_payments = pd.read_csv('/content/olist_order_payments_dataset.csv')
df_payments

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
0                          8          99.33  
1                          1          24.39  
2                          1          65.71  
3                          8         107.78  
4                          2         128.45  
...                      ...            ...  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58  

[103886 rows x 5 columns]

In [ ]:
df_payments.dtypes

# we already know that we have the correct data types in this dataset

order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

# **2 - Our Plan**

We are going to study Olist's customers' behaviors while ordering on the e-commerce platform. In particular, we are going to dig into the average order size and the frequency of orders for customers.

In order to achieve this goal, we will first:

**Clean our Data and Setup our Work Environment**

*   Hypothesis
*   Cleaning and Exploration of the data
*   Any comments?

**Follow up with some visualization using Plotly**

*   Parameters
*   Convert to dataframes
*   Line Chart : peak of orders over time
*   Bar Charts : customer behaviors regarding payment types and installment habits

**Perform our Statistical Analysis**

*   Mean
*   Median
*   Mode
*   Variance

**Conclusions**












In [ ]:
# in intro, we will clarify the relations between the tables (PK, FK, type)
# order_id is primary key for df_orders
# orders_id and payment_sequential are primary keys for df_orders_payments

In [ ]:
df_payments.keys()

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')

In [ ]:
df_orders_dataset.keys()

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

#  Cleaning Data and Preliminary Manipulations

# *Missing data*

In [ ]:
# Check missing columns orders_dataset
my_orders = df_orders_dataset.isnull().sum()
my_orders

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [ ]:
mv_payments = df_payments.isnull().sum()
mv_payments

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

# *Duplicates*

In [ ]:
# now we check for duplicates, first in df_orders_dataset

columns_to_check = ['order_id']

duplicate_orders_dataset = df_orders_dataset[columns_to_check].duplicated().any()
duplicate_orders_dataset

#other columns are not important and will have possible dublicates

# there are no duplicates, confirming that order_id is PK for this dataset



False

In [ ]:
# we check the duplicates in df_payments next

columns_to_check = ['order_id']

duplicate_payments = df_payments[columns_to_check].duplicated().any()
duplicate_payments

# there are duplicates in this df_orders_payments

True

# *Data Types*

In [ ]:
# We have to take care of converting the date columns in df_orders_dataset into date format
# ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

date_columns = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

for date in date_columns:
    df_orders_dataset[date] = pd.to_datetime(df_orders_dataset[date], format='%Y-%m-%d', errors='coerce')

df_orders_dataset.dtypes

order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

We observe some weird values that are not similar to the rest - e.g some timestamps are not dates for instance, some order_ids 99441 seem out of place too, etc.

In [ ]:
summary_stats_orders_dataset = df_orders_dataset.describe()

summary_stats_orders_dataset

<ipython-input-346-f1c9ffd0b094>:1: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.

<ipython-input-346-f1c9ffd0b094>:1: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.

<ipython-input-346-f1c9ffd0b094>:1: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.

<ipython-input-346-f1c9ffd0b094>:1: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and wi

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   
first                                NaN                               NaN   
last                                 NaN                               NaN   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   
first           NaN      2016-09-04 21:15:19  2016-09-15 12:16:38   
last            NaN      2018-10-17 17:30:18  2018-09-03 17:40:06   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   
first           2016-10-08 10:34:01           2016-10-11 13:46:32   
last            2018-09-11 19:48:28           2018-10-17 13:22:46   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522  
first            2016-09-30 00:00:00  
last             2018-11-12 00:00:00

In [ ]:
summary_payments = df_payments.describe()
summary_payments
# potentially round the results to 2 or 3 decimals maximum to clear the rows a bit.

payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

# *Replace Values and Join*

In [ ]:
# df_payments does not show any null values, which saves us time
# however, df_orders_dataset does contain NaN values - indeed, depending on the order status, not all the cells in the date/related columns can be filled - e.g if an order is placed but not yet validated, etc.

date_columns = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

for value in date_columns:
  if value == 'NaN':
    df_orders_dataset[date] = pd.replace(df_orders_dataset[date], value = 0)


In [ ]:
date_columns = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
df_orders_dataset[date_columns] = df_orders_dataset[date_columns].fillna(0)

In [ ]:
my_orders = df_orders_dataset.isnull().sum()
my_orders

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

Great! now our data is clean.
We can join our 2 datasets using the .merge() function.

In [ ]:
# it is also the equivalent of the Sales tables
# therefore we will perfom a join FROM df_orders_dataset LEFT JOIN df_orders_payments ON order_id

df_full_orders = pd.merge(df_orders_payments, df_orders_dataset, how ='left', on ='order_id')

df_full_orders

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value                       customer_id  \
0                          8          99.33  0a8556ac6be836b46b3e89920d59291c   
1                          1          24.39  f2c7fc58a9de810828715166c672f10a   
2                          1          65.71  25b14b69de0b6e184ae6fe2755e478f9   
3                          8         107.78  7a5d8efaaa1081f800628c30d2b0728f   
4                          2         128.45  15fd6fb8f8312dbb4674e4518d6fa3b3   
...                      ...            ...                               ...   
103881                     1         363.31  5d576cb2dfa3bc05612c392a1ee9c654   
103882                     2          96.80  2079230c765a88530822a34a4cec2aa0   
103883                     1          47.77  e4abb5057ec8cfda9759c0dc415a8188   
103884                     5         369.54  5d719b0d300663188169c6560e243f27   
103885                     1         191.58  4c7f868f43b5cff577b0becb8c8b7860   

       order_status order_purchase_timestamp    order_approved_at  \
0         delivered      2018-04-25 22:01:49  2018-04-25 22:15:09   
1         delivered      2018-06-26 11:01:38  2018-06-26 11:18:58   
2         delivered      2017-12-12 11:19:55  2017-12-14 09:52:34   
3         delivered      2017-12-06 12:04:06  2017-12-06 12:13:20   
4         delivered      2018-05-21 13:59:17  2018-05-21 16:14:41   
...             ...                      ...                  ...   
103881    delivered      2018-03-08 16:57:23  2018-03-10 03:55:25   
103882    delivered      2017-08-18 09:45:35  2017-08-18 10:04:56   
103883     invoiced      2017-11-18 17:27:14  2017-11-18 17:46:05   
103884    delivered      2018-08-07 23:26:13  2018-08-07 23:45:00   
103885    delivered      2017-12-05 08:18:35  2017-12-06 02:50:09   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2018-05-02 15:20:00           2018-05-09 17:36:51   
1               2018-06-28 14:18:00           2018-06-29 20:32:09   
2               2017-12-15 20:13:22           2017-12-18 17:24:41   
3               2017-12-07 20:28:28           2017-12-21 01:35:51   
4               2018-05-22 11:46:00           2018-06-01 21:44:53   
...                             ...                           ...   
103881          2018-03-12 18:19:36           2018-03-16 13:09:51   
103882          2017-08-18 18:04:24           2017-08-23 22:25:56   
103883                            0                             0   
103884          2018-08-09 11:46:00           2018-08-21 22:41:46   
103885          2017-12-08 00:51:47           2018-02-05 14:59:09   

       order_estimated_delivery_date  
0                         2018-05-22  
1                         2018-07-16  
2                         2018-01-04  
3                         2018-01-04  
4                         2018-06-13  
...                              ...  
103881                    2018-04-04  
103882                    2017-09-12  
103883                    2017-12-05  
103884                    2018-08-24  
103885                    2017-12-29  

[103886

Alright, the left join is now performed. Let's just make sure everything is in order and tidy before we dive into the analysis.

In [ ]:
full_orders = df_full_orders.isnull().sum()
full_orders

order_id                         0
payment_sequential               0
payment_type                     0
payment_installments             0
payment_value                    0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [ ]:
df_full_orders['order_id'].count()

103886

Perfect, now our data is ready to fuel our future dashboard.
Let's build nice charts in the next chapter of this project.

# **4 - Visualization**

Structure

1.   Order and Payment Trends Over Time

*   ID our peak periods, seasonality, special days
*   Can we already cluster our customers?

2.   Payment Installments Analysis

*   Can we come up with customer personas when it comes to using installment   options and payment options?
*   we use the installments, type and value columns

3.  Customer Retention

*   Analyze multiple purchase over time
*   Identify the best custumers - is there a correlation with some of the clusters mentioned above?

4.  Average Order Value (AOV)

*   Calculation
*   Are Olist's customers buying a lot and occasionnaly, or are they ordering often but smaller baskets.











In [ ]:
# let us first create a new figure to be called 'fig'

fig = go.Figure()

In [ ]:
# we want to study the time duration of our dataset in order to adjust the time unit of our charts

min_date_purchase = df_full_orders['order_purchase_timestamp'].min()
min_date_purchase

#max_value = df_full_orders['order_purchase_timestamp'].max()


Timestamp('2016-09-04 21:15:19')

In [ ]:
max_date_purchase = df_full_orders['order_purchase_timestamp'].max()
max_date_purchase


# 24 months time period - therefore let's consider a quarterly analysis of order behaviors

Timestamp('2018-10-17 17:30:18')

In [ ]:
# let's create a new column in our full dataset with a new time unit

df_full_orders['quarter'] = df_full_orders['order_purchase_timestamp'].dt.to_period('Q')

df_full_orders

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value                       customer_id  \
0                          8          99.33  0a8556ac6be836b46b3e89920d59291c   
1                          1          24.39  f2c7fc58a9de810828715166c672f10a   
2                          1          65.71  25b14b69de0b6e184ae6fe2755e478f9   
3                          8         107.78  7a5d8efaaa1081f800628c30d2b0728f   
4                          2         128.45  15fd6fb8f8312dbb4674e4518d6fa3b3   
...                      ...            ...                               ...   
103881                     1         363.31  5d576cb2dfa3bc05612c392a1ee9c654   
103882                     2          96.80  2079230c765a88530822a34a4cec2aa0   
103883                     1          47.77  e4abb5057ec8cfda9759c0dc415a8188   
103884                     5         369.54  5d719b0d300663188169c6560e243f27   
103885                     1         191.58  4c7f868f43b5cff577b0becb8c8b7860   

       order_status order_purchase_timestamp    order_approved_at  \
0         delivered      2018-04-25 22:01:49  2018-04-25 22:15:09   
1         delivered      2018-06-26 11:01:38  2018-06-26 11:18:58   
2         delivered      2017-12-12 11:19:55  2017-12-14 09:52:34   
3         delivered      2017-12-06 12:04:06  2017-12-06 12:13:20   
4         delivered      2018-05-21 13:59:17  2018-05-21 16:14:41   
...             ...                      ...                  ...   
103881    delivered      2018-03-08 16:57:23  2018-03-10 03:55:25   
103882    delivered      2017-08-18 09:45:35  2017-08-18 10:04:56   
103883     invoiced      2017-11-18 17:27:14  2017-11-18 17:46:05   
103884    delivered      2018-08-07 23:26:13  2018-08-07 23:45:00   
103885    delivered      2017-12-05 08:18:35  2017-12-06 02:50:09   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2018-05-02 15:20:00           2018-05-09 17:36:51   
1               2018-06-28 14:18:00           2018-06-29 20:32:09   
2               2017-12-15 20:13:22           2017-12-18 17:24:41   
3               2017-12-07 20:28:28           2017-12-21 01:35:51   
4               2018-05-22 11:46:00           2018-06-01 21:44:53   
...                             ...                           ...   
103881          2018-03-12 18:19:36           2018-03-16 13:09:51   
103882          2017-08-18 18:04:24           2017-08-23 22:25:56   
103883                            0                             0   
103884          2018-08-09 11:46:00           2018-08-21 22:41:46   
103885          2017-12-08 00:51:47           2018-02-05 14:59:09   

       order_estimated_delivery_date quarter  
0                         2018-05-22  2018Q2  
1                         2018-07-16  2018Q2  
2                         2018-01-04  2017Q4  
3                         2018-01-04  2017Q4  
4                         2018-06-13  2018Q2  
...                              ...     ...  
103881                    2018-04-04  2018Q1  
103882                    2017-09-12  2017Q3  
103883                    2017-12-05  2017Q4  
103884

In [ ]:
# we then need to aggregate/group our data by this new time unit

quarterly_orders = df_full_orders.groupby(['quarter'])['order_id'].count()

quarterly_orders



quarter
2016Q3        3
2016Q4      343
2017Q1     5573
2017Q2     9951
2017Q3    13383
2017Q4    18618
2018Q1    22027
2018Q2    20763
2018Q3    13221
2018Q4        4
Freq: Q-DEC, Name: order_id, dtype: int64

In [ ]:
# Dataframe convertion for plotting

df_quarterly = quarterly_orders.to_frame().reset_index()
df_quarterly

quarter  order_id
0  2016Q3         3
1  2016Q4       343
2  2017Q1      5573
3  2017Q2      9951
4  2017Q3     13383
5  2017Q4     18618
6  2018Q1     22027
7  2018Q2     20763
8  2018Q3     13221
9  2018Q4         4

In [ ]:
df_quarterly.dtypes

quarter     period[Q-DEC]
order_id            int64
dtype: object

In [ ]:
df_quarterly['quarter'] = df_quarterly['quarter'].astype(str)

In [ ]:
# a line chart will show us the potential peaks of orders throughout the entire period

fig = px.line(df_quarterly, x='quarter', y='order_id', labels={'order_id': 'Amount of orders', 'quarter': 'Quarter of Year'}, title= 'A peak in the order numbers in Q1 2018')
fig.show()



In [ ]:
# now Visualization: Customer Segmentation Scatter Plot Over Time
# we need to categorize our customers payments habits and type
# let's group by : payment types and installents

In [ ]:
payment_types = df_full_orders.groupby(['payment_type'])['order_id'].count()

payment_types


payment_type
boleto         19784
credit_card    76795
debit_card      1529
not_defined        3
voucher         5775
Name: order_id, dtype: int64

In [ ]:
df_payment_types = payment_types.to_frame().reset_index()
df_payment_types

payment_type  order_id
0       boleto     19784
1  credit_card     76795
2   debit_card      1529
3  not_defined         3
4      voucher      5775

In [ ]:
# which is the preferred method of payment?
# bar chart

fig = px.bar(df_payment_types, x='payment_type', y='order_id', title= 'The Credit Card game is strong')

fig.show()

# a pie chart could have worked too

In [ ]:
# lets do the same but for installments and customer behaviorial

installments = df_full_orders.groupby(['payment_installments'])['order_id'].count()

installments


payment_installments
0         2
1     52546
2     12413
3     10461
4      7098
5      5239
6      3920
7      1626
8      4268
9       644
10     5328
11       23
12      133
13       16
14       15
15       74
16        5
17        8
18       27
20       17
21        3
22        1
23        1
24       18
Name: order_id, dtype: int64

In [ ]:
df_installments = installments.to_frame().reset_index()
df_installments

payment_installments  order_id
0                      0         2
1                      1     52546
2                      2     12413
3                      3     10461
4                      4      7098
5                      5      5239
6                      6      3920
7                      7      1626
8                      8      4268
9                      9       644
10                    10      5328
11                    11        23
12                    12       133
13                    13        16
14                    14        15
15                    15        74
16                    16         5
17                    17         8
18                    18        27
19                    20        17
20                    21         3
21                    22         1
22                    23         1
23                    24        18

In [ ]:
# bar chart

fig = px.bar(df_installments, x='payment_installments', y='order_id', title= 'Most customers pay in 1 to 3 installments')

fig.show()

In [ ]:
fig.update_layout(xaxis=dict(range=[0,13]))

In [ ]:
# we have 2 valid dataframes

df_merged_scatter = pd.merge(df_payment_types, df_installments, how='right', on='order_id')

df_merged_scatter

payment_type  order_id  payment_installments
0           NaN         2                     0
1           NaN     52546                     1
2           NaN     12413                     2
3           NaN     10461                     3
4           NaN      7098                     4
5           NaN      5239                     5
6           NaN      3920                     6
7           NaN      1626                     7
8           NaN      4268                     8
9           NaN       644                     9
10          NaN      5328                    10
11          NaN        23                    11
12          NaN       133                    12
13          NaN        16                    13
14          NaN        15                    14
15          NaN        74                    15
16          NaN         5                    16
17          NaN         8                    17
18          NaN        27                    18
19          NaN        17                    20
20  not_defined         3                    21
21          NaN         1                    22
22          NaN         1                    23
23          NaN        18                    24

In [ ]:
# the following were our ambitions for the visualisation part

Payment Installments Behavior:

Visualization: Use a bar chart or pie chart to show the distribution of orders with and without installments. You can also create a line chart to display the trend in the usage of installments over time.
Library: Matplotlib or Seaborn.
Correlation Between Type, Installments, and Sequential:

Visualization: Create a heatmap to visualize the correlation matrix between these variables.
Library: Seaborn or Matplotlib for the heatmap.
Relation Between Type of Payments and Installments:

Visualization: Stacked bar chart or grouped bar chart to show the distribution of payment types for orders with and without installments.
Library: Matplotlib or Seaborn.
Customer Categorization Based on Average Order Value:

Visualization: Use a scatter plot or box plot to visualize the distribution of average order values for different customer segments.
Library: Matplotlib or Seaborn.
Preferred Ordering Time Analysis:

Visualization: Line chart or bar chart to show the count of orders over different time intervals (e.g., hours of the day, days of the week).
Library: Matplotlib or Seaborn.
Delivery Process Time Analysis:

Visualization: Create a box plot or violin plot to display the distribution of delivery process times (time to validate, time to ship, time to deliver).
Library: Matplotlib or Seaborn.

# **5 - Statistical tests**

We want to analyse the correlation in between:
- payment type and payment installment habits
- payment installment habits and payment value

Are there obvious conclusions to be drawn?
We are already assuming that the higher the value, the higher the probability to pay by installments (BUT not a necessity). Let's test that.

Also, we make the hypothesis that there is a positive correlation between the type of payment and the number of installments tied to an order - e.g voucher = 1 payment, credit card = several, etc.

In [ ]:
# our full table is df_full_orders

df_full_orders


order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value                       customer_id  \
0                          8          99.33  0a8556ac6be836b46b3e89920d59291c   
1                          1          24.39  f2c7fc58a9de810828715166c672f10a   
2                          1          65.71  25b14b69de0b6e184ae6fe2755e478f9   
3                          8         107.78  7a5d8efaaa1081f800628c30d2b0728f   
4                          2         128.45  15fd6fb8f8312dbb4674e4518d6fa3b3   
...                      ...            ...                               ...   
103881                     1         363.31  5d576cb2dfa3bc05612c392a1ee9c654   
103882                     2          96.80  2079230c765a88530822a34a4cec2aa0   
103883                     1          47.77  e4abb5057ec8cfda9759c0dc415a8188   
103884                     5         369.54  5d719b0d300663188169c6560e243f27   
103885                     1         191.58  4c7f868f43b5cff577b0becb8c8b7860   

       order_status order_purchase_timestamp    order_approved_at  \
0         delivered      2018-04-25 22:01:49  2018-04-25 22:15:09   
1         delivered      2018-06-26 11:01:38  2018-06-26 11:18:58   
2         delivered      2017-12-12 11:19:55  2017-12-14 09:52:34   
3         delivered      2017-12-06 12:04:06  2017-12-06 12:13:20   
4         delivered      2018-05-21 13:59:17  2018-05-21 16:14:41   
...             ...                      ...                  ...   
103881    delivered      2018-03-08 16:57:23  2018-03-10 03:55:25   
103882    delivered      2017-08-18 09:45:35  2017-08-18 10:04:56   
103883     invoiced      2017-11-18 17:27:14  2017-11-18 17:46:05   
103884    delivered      2018-08-07 23:26:13  2018-08-07 23:45:00   
103885    delivered      2017-12-05 08:18:35  2017-12-06 02:50:09   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2018-05-02 15:20:00           2018-05-09 17:36:51   
1               2018-06-28 14:18:00           2018-06-29 20:32:09   
2               2017-12-15 20:13:22           2017-12-18 17:24:41   
3               2017-12-07 20:28:28           2017-12-21 01:35:51   
4               2018-05-22 11:46:00           2018-06-01 21:44:53   
...                             ...                           ...   
103881          2018-03-12 18:19:36           2018-03-16 13:09:51   
103882          2017-08-18 18:04:24           2017-08-23 22:25:56   
103883                            0                             0   
103884          2018-08-09 11:46:00           2018-08-21 22:41:46   
103885          2017-12-08 00:51:47           2018-02-05 14:59:09   

       order_estimated_delivery_date quarter  
0                         2018-05-22  2018Q2  
1                         2018-07-16  2018Q2  
2                         2018-01-04  2017Q4  
3                         2018-01-04  2017Q4  
4                         2018-06-13  2018Q2  
...                              ...     ...  
103881                    2018-04-04  2018Q1  
103882                    2017-09-12  2017Q3  
103883                    2017-12-05  2017Q4  
103884

# *Correlation 1*

In [ ]:
import pandas as pd
import plotly.graph_objects as go

# We create a contingency table
contingency_table = pd.crosstab(df_full_orders['payment_type'], df_full_orders['payment_installments'])

# Now a heatmap will certainly give us insights on how customers pay and how long they usually need to pay in full

heatmap = go.Figure(data=go.Heatmap(z=contingency_table.values,
                                    x=contingency_table.columns,
                                    y=contingency_table.index,
                                    colorscale='blues'))  # Adjust the colorscale

# Plot labels and title
heatmap.update_layout(title='Correlation Heatmap: Payment Type vs Installments',
                      xaxis_title='Installments',
                      yaxis_title='Payment Type')
heatmap.show()

In [ ]:
# payment installment habits and payment value

# *Correlation 2*

In [ ]:
df_full_orders['payment_installments'].corr(df_full_orders['payment_value'])

0.33081084451898546

# 6 - **Conclusions**

Before we dive into observations, we wanted to explore additional visualizations or conduct further analysis - e.g examining the distribution of payment values for different installment options or considering other factors that might influence payment behavior.


**Correlation 1: relation between the type of payment and the installment habits**

>> There is only some correlation between credit card and installment habits.
NB: voucher it makes sense. Boleto Bancário, commonly known as Boleto, is a popular Brazilian cash-based payment method with 15% market share. It is also the only payment method available for the 17% of Brazilian shoppers who don't have a bank account = logical.
we should ask the Olist finance team more info about the 'not defined'.
You must pay by full with a debit card.

>> The Credit Card is definitely the most used type of payment.

>> Brazilians favorite payment method is still domestic credit cards, which are accepted by 98.3% of online retailers. For consumers, the main advantage of credit cards is the option of paying in monthly installments, which is a demand very well understood by virtual stores,  with 32.2% of Brazilian e-commerce platforms offering payment in up to ten monthly installments, which is the most common payment plan offered in the country.


**Correlation 2: relation between the number of installments and the order value**

>> The correlation coefficient between 'payment_installments' and 'payment_value' is approximately 0.33.

>> Positive Correlation: The correlation coefficient is positive, indicating a positive relationship between the two variables. As one variable increases, the other tends to increase as well.

>> Moderate: it indicates that there is a tendency for the number of payment installments to increase as the payment value increases, and vice versa.

>> Warning: not causation. we can observe a statistical association between these variables, but we cannot conclude that one variable causes the other to change.

